server/modlog.ts

```diff
@@ -36,284 +32,279 @@ const PUNISHMENTS = [
+		if (oldID === newID) return;
+		void this.runSQL({statement: this.renameQuery, args: [newID, oldID]});
 	}
 
-	async destroyAll() {
-		const promises = [];
-		for (const id in this.streams) {
-			promises.push(this.destroy(id as ModlogID));
-		}
-		return Promise.all(promises);
+	/******************************************
+	 * Methods for reading (searching) modlog *
+	 ******************************************/
+	generateRegex(search: string, isExact?: boolean) {
+		// Ensure the generated regex can never be greater than or equal to the value of
+		// RegExpMacroAssembler::kMaxRegister in v8 (currently 1 << 16 - 1) given a
+		// search with max length MAX_QUERY_LENGTH. Otherwise, the modlog
+		// child process will crash when attempting to execute any RegExp
+		// constructed with it (i.e. when not configured to use ripgrep).
+		if (isExact) return search.replace(/[\\.+*?()|[\]{}^$]/g, '\\$&');
+		return `[^a-zA-Z0-9]?${[...search].join('[^a-zA-Z0-9]*')}([^a-zA-Z0-9]|\\z)`;
 	}
 
-	async rename(oldID: ModlogID, newID: ModlogID) {
-		const streamExists = this.streams.has(oldID);
-		if (streamExists) await this.destroy(oldID);
-		await FS(`${this.logPath}/modlog_${oldID}.txt`).rename(`${this.logPath}/modlog_${newID}.txt`);
-		if (streamExists) this.initialize(newID);
+	formatArray(arr: unknown[], args: unknown[]) {
+		args.push(...arr);
+		return [...'?'.repeat(arr.length)].join(', ');
 	}
 
-	getActiveStreamIDs() {
-		return [...this.streams.keys()];
-	}
+	prepareSearch(
+		rooms: ModlogID[], maxLines: number, onlyPunishments: boolean, search: ModlogSearch
+	): ModlogQuery<string | number> {
+		for (const room of [...rooms]) {
+			rooms.push(`global-${room}` as ModlogID);
+		}
+		const userid = toID(search.user);
 
-	/******************************************
-	 * Methods for reading (searching) modlog *
-	 ******************************************/
-	 async runSearch(
-		rooms: ModlogID[], search: string, isExact: boolean, maxLines: number, onlyPunishments: boolean | string
-	) {
-		const useRipgrep = await checkRipgrepAvailability();
-		let fileNameList: string[] = [];
-		let checkAllRooms = false;
-		for (const roomid of rooms) {
-			if (roomid === 'all') {
-				checkAllRooms = true;
-				const fileList = await FS(this.logPath).readdir();
-				for (const file of fileList) {
-					if (file !== 'README.md' && file !== 'modlog_global.txt') fileNameList.push(file);
-				}
-			} else {
-				fileNameList.push(`modlog_${roomid}.txt`);
-			}
+		const args: (string | number)[] = [];
+
+		let roomChecker = `roomid IN (${this.formatArray(rooms, args)})`;
+		if (rooms.includes('global')) roomChecker = `(roomid LIKE 'global-%' OR ${roomChecker})`;
+
+		let query = `SELECT *, (SELECT group_concat(userid, ',') FROM alts WHERE alts.modlog_id = modlog.modlog_id) as alts FROM modlog`;
+		query += ` WHERE ${roomChecker}`;
+
+		if (search.action) {
+			query += ` AND action LIKE '%' || ? || '%'`;
+			args.push(search.action.toUpperCase());
```

Comment:
Would rather make it an invariant of `search.action` to be uppercase (iow, when you create a `ModlogSearch`, ensure it is already uppercase)